Upgrade tokio to 1.22.0 and replace async-std with tokio

dmitry-markin opened the pull request:
Resolves https://github.com/paritytech/substrate/issues/12499.

polkadot companion: https://github.com/paritytech/polkadot/pull/6262

cumulus companion: https://github.com/paritytech/cumulus/pull/1917

altonen reviewed (commented):
Overall LGTM but I have two questions:

* is there a reason why `async_std::spawn` in tests is not converted to `tokio::spawn`?
* the example code in `rust-libp2p` repository instructs to configure executor for the swarm in [tokio chat example](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L112). Should we configure the default executor similarly in case `params.executor` is `None` in [`service.rs`](https://github.com/paritytech/substrate/blob/master/client/network/src/service.rs#L396-L404)?

I'm paranoid about the second bullet because I burned two weeks debugging bizarre async execution issues after having configured libp2p incorrectly.

koute commented:
Nice! :+1: 

We should completely nuke `async-std` so that it isn't used anywhere in our codebases. Doesn't really make sense to have two different executors, especially when `tokio`'s solid.

bkchr commented on bin/node/cli/tests/websocket_server.rs:
```diff
@@ -16,7 +16,8 @@
 
-use async_std::net::{TcpListener, TcpStream};
+use tokio::net::{TcpListener, TcpStream};
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt};
```
Why do need compatibility mode?

dmitry-markin replied:
This is because soketto WebSocket server expects `TcpStream` to be `futures::io::AsyncRead` and not `tokio::io::AsyncRead`, so we need to wrap tokio's `TcpStream` with a `.compat()` call here: https://github.com/paritytech/substrate/blob/398d10c3ee6d8f0c48acbb0f81401d93fb25245f/bin/node/cli/tests/websocket_server.rs#L128

Surprisingly, I also haven't found other alternative with libp2p protocol upgrades, which also expect `futures::io::AsyncRead`, so tokio's `TcpStream`s are also wrapped in `.compat()` there. Please correct me if there is a better way around. https://github.com/paritytech/substrate/blob/398d10c3ee6d8f0c48acbb0f81401d93fb25245f/client/network/src/protocol/notifications/upgrade/notifications.rs#L501-L505

bkchr replied:
Ty for the explanation! I should have checked before. I thought that `Compat` is for making it compatible for older tokio versions. There was once a similar construct for using futures v0.3 and v0.1 together ;) That confused me here. But it sounds good what you are doing here!

dmitry-markin commented:
> * the example code in `rust-libp2p` repository instructs to configure executor for the swarm in [tokio chat example](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat-tokio.rs#L112). Should we configure the default executor similarly in case `params.executor` is `None` in [`service.rs`](https://github.com/paritytech/substrate/blob/master/client/network/src/service.rs#L396-L404)?
> 
> 
> I'm paranoid about the second bullet because I burned two weeks debugging bizarre async execution issues after having configured libp2p incorrectly.

Interestingly, the PR https://github.com/libp2p/rust-libp2p/pull/3097 has just landed in libp2p that makes the specification of executor explicit during the construction of `Swarm`: `SwarmBuilder::new` has been replaced with `SwarmBuilder::with_executor` + some alternative constructors with predifined executors for tokio, async-std, etc.

I'm going to change the API of `NetworkWorker::new` so it won't accept `None` executor anymore in this PR, and later when libp2p-0.50 is released we can move to the new `SwarmBuilder::with_executor` constructor.

CC @altonen

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2082447

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2082433

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2083153

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2083164

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2087632

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2090980

dmitry-markin commented:
I suspect that I have converted some `futures::executor::block_on` into `runtime.block_on` where it wasn't strictly necessary. I also don't like that in some tests we create tokio runtime explicitly and split sync initialization and async execution, and in others we just run everything under `#[tokio::test]`. So, may be a follow-up PR/commits are needed to clean these things up @altonen @bkchr.

melekes commented:
libp2p 0.50 was released last week. here is the PR that updates substrate: https://github.com/paritytech/substrate/pull/12734

let me know if you want me to change executors to tokio.

melekes commented:
libp2p 0.50 was released last week. here is the PR that updates substrate: https://github.com/paritytech/substrate/pull/12734

let me know if you want me to change executors to tokio.

melekes commented on client/network/src/service.rs:
```diff
@@ -383,15 +383,15 @@ where
-						(self.0)(f)
-					}
+
+			struct SpawnImpl<F>(F);
```
what are the benefits of this executor versus say Tokio's one?

melekes reviewed (commented):
👍

koute commented on client/network/src/config.rs:
```diff
@@ -68,7 +68,7 @@ where
 	/// How to spawn background tasks. If you pass `None`, then a threads pool will be used by
 	/// default.
-	pub executor: Option<Box<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>) + Send>>,
+	pub executor: Box<dyn Fn(Pin<Box<dyn Future<Output = ()> + Send>>) + Send>,
```
Looks like the comment needs to be updated here.

dmitry-markin replied:
Sure, thanks!

dmitry-markin commented on client/beefy/src/tests.rs:
```diff
@@ -713,13 +726,15 @@ fn correct_beefy_payload() {
 	sp_tracing::try_init_simple();
 
-	let mut net = BeefyTestNet::new(2);
+	let runtime = Runtime::new().unwrap();
```
We can. I think, historically some tests were synchronous and later converted to async ones with manual runtime instantiation, and I just repeated the mistake and did the same in other places. Probably this doesn't make sense, so I'm going to convert all these tests to `#[tokio::test]` in a follow-up PR.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2113708

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2113704

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2113703

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2113709

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2113710

dmitry-markin commented:
bot merge
